File: Data Structures & Algorithms/binary-tree-diameter/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    // max of left subtree depth + right subtree depth
    int res = 0;
    int helper(TreeNode* root)
    {
        if(root == nullptr) return 0;

        int left = helper(root->left);
        int right = helper(root->right);

        // set res to left depth + right depth + 1
        res = max(res, left + right + 1);

        // returns max depth
        return max(left, right) + 1;
    }

    int diameterOfBinaryTree(TreeNode* root) {
        helper(root);
        return res - 1;
    }
};

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def helper(self, root: Optional[TreeNode]) -> Tuple[int, int]:
        if root is None:
            return [-1001, -1001]
        left, left_root = self.helper(root.left)
        right, right_root = self.helper(root.right)
        res = max(left, right, root.val + max(0, left_root) + max(0, right_root))
        max_path = root.val + max(left_root, right_root, 0)
        return [res, max_path]

    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        return self.helper(root)[0]